Allow prop.createSent to spawn more SENTs. closes #919

DaDamRival opened the pull request:
This isn't finished yet, still gotta add more SENTs to the list and maby even some Wiremod ones.

Example
```lua
--@server

prop.createSent(chip():getPos(), chip():getAngles(), "gmod_button", true, {
    key = 39, -- numpad 2, gotta make the KEY enum shared
    label = "test"
})
```

thegrb93 commented:
Looks like a good start. I think class definitions ought to go into a new file. Another problem will be documenting the variable classes' data structures. Perhaps just provide a link to the lua file.

DaDamRival commented:
Added more default Gmod sents and simple docs by using the table docs class in the prop library. Another option would be to make a separate library for all sents and every entry in it would be a separate data structure.

Also moved the structures to a separate file but no clue where it should be located so put it in libs_sv/ for now.

adamnejm commented:
Would be nice if user didn't have to provide an empty table if he's fine with default parameters.

thegrb93 commented on lua/starfall/libs_sv/prop_sent.lua:
```diff
@@ -0,0 +1,1595 @@
+-- Simple helper function that provides model if it doesnt exist
+local function registerSent(class, data)
+	data[1].model = data[1].model or {"Model", TYPE_STRING, "models/props_junk/watermelon01.mdl"}
```
`if not data[1].model then data[1].model = {"Model", TYPE_STRING, "models/props_junk/watermelon01.mdl"} end`

thegrb93 commented on lua/starfall/libs_sv/prop.lua:
```diff
@@ -430,9 +458,48 @@ function props_library.createSent(pos, ang, class, frozen)
+			local value = data[param]
+			
+			if value then
+				checkluatype(value, org[2])
```
Better error message here will be needed. Maybe check type TypeID(value)==org[2] and if not then SF.ThrowTypeError https://github.com/thegrb93/StarfallEx/blob/master/lua/starfall/libs_sh/builtins.lua#L756

thegrb93 commented on lua/starfall/libs_sv/prop.lua:
```diff
@@ -430,9 +458,48 @@ function props_library.createSent(pos, ang, class, frozen)
+		end
+		
+		-- Supply additional data
+		enttbl.Data = enttbl
```
This should be two different tables. The assignments below may overwrite some data.

DaDamRival replied:
Data is a special case that passes the whole structure.
https://wiki.facepunch.com/gmod/duplicator.RegisterEntityClass

thegrb93 replied:
Ok. Weird

thegrb93 commented on lua/starfall/libs_sv/prop_sent.lua:
```diff
@@ -0,0 +1,1595 @@
+	list.Set("starfall_creatable_sent", class, data)
+end
+
+local function tryCastColor(tbl, allow_alpha)
```
Id just name this castColor. 'try' implies that it could fail and this implementation will never fail.

thegrb93 commented:
TYPE_TABLE data will need additional security I think.

DaDamRival commented:
TYPE_TABLE is only used for colors only (except for gmod_wire_textreceiver but that checks the tables contents) since there is no TYPE_COLOR.

thegrb93 commented on lua/starfall/libs_sv/prop.lua:
```diff
@@ -472,7 +472,19 @@ function props_library.createSent(pos, ang, class, frozen, data)
+				-- checkluatype(value, org[2])
+				local typeid = TypeID(value)
+				if typeid ~= org[2] then
+					local typeLookup = {
```
I think using SF.GetType here and also modifying the SF.ThrowTypeError to take prefix text would be better than rewriting it all here.

thegrb93 replied:
Also if we can make a type lookup function SF.TypeName that does this conversion. https://github.com/thegrb93/StarfallEx/blob/master/lua/starfall/sflib.lua#L648

DaDamRival replied:
Made the changes to sflib a separate pr #923

thegrb93 replied:
Put it in this PR instead. I don't want to complicate things.

Damianu commented on lua/starfall/sflib.lua:
```diff
@@ -628,14 +628,14 @@ function SF.Throw(msg, level, uncatchable)
 
 --- Throws a type error
+-- @param prefix Prefix to add before 'in function'
 -- @param expected The exprect type name
```
👎

thegrb93 commented on lua/starfall/libs_sv/prop_sent.lua:
```diff
@@ -0,0 +1,1702 @@
+}})
+
+-- Chip bois
+registerSent("gmod_wire_expression2", {
```
Wow lol, starfall making e2 its bih

thegrb93 commented on lua/starfall/libs_sv/prop_sent.lua:
```diff
@@ -992,7 +1040,36 @@ registerSent("gmod_wire_adv_emarker", {{
+		pos           = {"LPos",       TYPE_VECTOR, Vector()},
+		axis          = {"LAxis",      TYPE_VECTOR, Vector(0, 1, 0)},
+		forward_value = {"fwd",        TYPE_NUMBER, 1},
+		back_value    = {"bck",        TYPE_NUMBER, -1},
+		stop_value    = {"bck",        TYPE_NUMBER, 0},
```
Are these supposed to both be "bck"?

DaDamRival replied:
Woops, fixed

thegrb93 commented on lua/starfall/libs_sv/prop.lua:
```diff
@@ -472,7 +472,7 @@ function props_library.createSent(pos, ang, class, frozen, data)
 			
 			if value then
-				value = ounwrap(value)
+				value = ounwrap(value) or value
```
What value is ounwrap returning nil? You shouldn't need this.

DaDamRival replied:
Table, but looking at the sourcecode for UnwrapObject TYPE_TABLE could probably be added to safe_types.

thegrb93 replied:
Ah ok that makes sense. TYPE_TABLE can't be added to safe types. ounwrap might be able to return them, but I have to first check that there's no security problems with that.

thegrb93 replied:
I see, no, tables have to be explicitly allowed so the code here is correct.

thegrb93 commented on lua/starfall/libs_sv/prop.lua:
```diff
@@ -279,15 +278,46 @@ function props_library.createComponent(pos, ang, class, model, frozen)
 end
 
+--- Get a list of all spawnable sents.
+-- @param ordered True to get an ordered list
```
I'd rename ordered to categorized

thegrb93 commented on lua/starfall/libs_sv/prop.lua:
```diff
@@ -279,15 +278,46 @@ function props_library.createComponent(pos, ang, class, model, frozen)
 -- @param ang Angle of created sent
 -- @param class Class of created sent
 -- @param frozen True to spawn frozen
+-- @param data Optional table, additional entity data to be supplied
```
I'd specify the data is for specific sents and direct the user to the documentation of those sents

thegrb93 commented on lua/starfall/libs_sv/prop.lua:
```diff
@@ -430,9 +458,57 @@ function props_library.createSent(pos, ang, class, frozen)
+		end
+		
+		if entity.PreEntityCopy then
+			entity:PreEntityCopy()
```
Why's this getting called?

DaDamRival replied:
Wiremod entities add a entry to the table copy which is used for PostEntityPaste. Without this e2 wouldn't execute.

thegrb93 commented on lua/starfall/libs_sv/prop.lua:
```diff
@@ -430,9 +458,57 @@ function props_library.createSent(pos, ang, class, frozen)
+				
+				enttbl[org[1]] = value
+			else
+				enttbl[org[1]] = org[3]
```
Is there always a default? I think an error would be better in some cases so users know they are misusing the SENT.

thegrb93 replied:
I've added ability for default to be missing. I'd recommend removing defaults that don't make sense so that the user must provide that data.

thegrb93 commented on lua/starfall/libs_sv/prop.lua:
```diff
@@ -430,9 +458,57 @@ function props_library.createSent(pos, ang, class, frozen)
+				
+				checkluatype(value, org[2], nil, "Parameter: " .. param)
+				
+				enttbl[org[1]] = value
```
We need to check if TYPE_TABLE and create a new table and copy the specific indeces if so. Users' tables can't be trusted.

thegrb93 replied:
It would also be good to detect invalid keys in 'data' and throw an error.

DaDamRival commented on lua/starfall/libs_sv/prop.lua:
```diff
@@ -481,9 +481,20 @@ function props_library.createSent(pos, ang, class, frozen, data)
 				
-				enttbl[org[1]] = value
-			else
+				if istable(value) then
```
Not all tables are a color, e2 included for example. This also wont allow for tables structured with rgb indexes, this is thje reason castColor exists.

thegrb93 commented:
Just need to remove defaults that don't make sense, make sure the doc gen still works with that, do docgen, make sure everything works.

thegrb93 commented:
I think it's working now. I've tested with a gmod_thruster. Might test a few more entities first.
